tianshou/data/buffer/base.py

```diff
@@ -290,20 +290,32 @@ def add(
             self._meta[ptr] = batch
         return ptr, ep_rew, ep_len, ep_idx
 
-    def sample_indices(self, batch_size: int) -> np.ndarray:
+    def sample_indices(self, batch_size: int | None) -> np.ndarray:
```

Comment:
This change is actually wrong and unnecessary... I stepped back to take a look at #340, it uses https://github.com/thu-ml/tianshou/blob/a57503c0aa6d40a0c939b37f4ae3c3a83bb2e126/tianshou/trainer/onpolicy.py#L134-L136 which has different semantics -- sample size is 0 which will extract all data from buffer, but batch_size is 99999 which will presumably update all data in one forward-backward call. You should revive that change and make some classes' policy.update function accept batch_size=None to allow such a feature, instead of changing here

Reply:
This change seems less innocent than I anticipated. I'd like to address it in a separate issue, maybe at the same time addressing the TODOs I added to buffer. I am a bit afraid of touching this code since it's at the core of everything. Would you have time to work on it or do some pair programming with me in the near future?

Meanwhile, I'd drop this commit here and open a new draft branch

Reply:
k will push here later